{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import os\n",
    "import numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bedgraph(bedgraph):\n",
    "    res = []\n",
    "    last = 0\n",
    "    for line in open(bedgraph):\n",
    "        tab = line.split()\n",
    "        end = int(tab[2])\n",
    "        for _ in range(last, end):\n",
    "            res.append(float(tab[3]))\n",
    "            last = end\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "early_R1_plus = read_bedgraph('Early-5h/S1-5h_5cov+.bedgraph')\n",
    "early_R1_minus = read_bedgraph('Early-5h/S1-5h_5cov-.bedgraph')\n",
    "early_R2_plus = read_bedgraph('Early-5h/S2-5h_5cov+.bedgraph')\n",
    "early_R2_minus = read_bedgraph('Early-5h/S2-5h_5cov-.bedgraph')\n",
    "late_R1_plus = read_bedgraph('Late-16h/S3-16h_5cov+.bedgraph')\n",
    "late_R1_minus = read_bedgraph('Late-16h/S3-16h_5cov-.bedgraph')\n",
    "late_R2_plus = read_bedgraph('Late-16h/S4-16h_5cov+.bedgraph')\n",
    "late_R2_minus = read_bedgraph('Late-16h/S4-16h_5cov-.bedgraph')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_peaks(vector, window_size=10, min_height=10):\n",
    "    peaks = []\n",
    "    for i in range(len(vector) - window_size + 1):\n",
    "        values = vector[i: i+window_size]\n",
    "        max_value = -1; max_pos = -1\n",
    "        for index,value in enumerate(values):\n",
    "            if value > max_value:\n",
    "                max_value = value\n",
    "                max_pos = i + index        \n",
    "        if max_value >= min_height:\n",
    "            if peaks:\n",
    "                prev_max, prev_pos = peaks[-1]\n",
    "                if max_pos - prev_pos >= window_size:\n",
    "                    peaks.append((max_value, max_pos))\n",
    "                else:\n",
    "                    if prev_max < max_value:\n",
    "                        peaks.pop()\n",
    "                        peaks.append((max_value, max_pos))\n",
    "            else:\n",
    "                peaks.append((max_value, max_pos))\n",
    "    return peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## check correspondence\n",
    "def check_correspondence(values1, values2, distance=10):\n",
    "    if len(values1) > len(values2):\n",
    "        values1, values2 = values2, values1\n",
    "    matching_v2 = []\n",
    "    for v2 in values2:\n",
    "        for v1 in values1:\n",
    "            if abs(v1[1] - v2[1]) <= distance:\n",
    "                matching_v2.append(v2)\n",
    "                break\n",
    "    return len(matching_v2) / len(values2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_genes(gff='Genome Info/U18466.2.gff3', out_bedfile='Genome Info/U18466.2_genes.bed'):\n",
    "    genes = []\n",
    "    with open(out_bedfile, 'w') as f:\n",
    "        for line in open(gff):\n",
    "            tab = line.split('\\t')\n",
    "            if len(tab) < 2:\n",
    "                continue\n",
    "            if tab[2] == 'gene':\n",
    "                start, end, strand = int(tab[3]), int(tab[4]), tab[6]\n",
    "                name = tab[8].split('Name=')[1].split(';')[0]\n",
    "                genes.append((start, end, strand, name))\n",
    "                print(tab[0], start-1, end-1, name, '.', tab[6], sep='\\t', file=f)\n",
    "    return genes\n",
    "\n",
    "genes = read_genes()       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "## get subset of genes with mapped start sites\n",
    "def get_genes_starts(genes, peaks):\n",
    "    tmp_start = 'tmp_tss.bed'\n",
    "    with open(tmp_start, 'w') as f:\n",
    "        peaks.sort(key=lambda x: x[0])\n",
    "        for start, strand, value in peaks:\n",
    "            print('U18466.2', start-1, start, '.', value, strand, sep='\\t', file=f)\n",
    "    cmd = 'bedtools closest -s -D a -io -iu -a \"%s\" -b \"%s\" > tmp' % (tmp_start, 'Genome Info/U18466.2_genes.bed')\n",
    "    os.system(cmd)\n",
    "    gene2tss = {}\n",
    "    for line in open('tmp'):\n",
    "        tab = line.split()\n",
    "        if tab[9] not in gene2tss:\n",
    "            gene2tss[tab[9]] = []\n",
    "        gene2tss[tab[9]].append((int(tab[2]), float(tab[4])))\n",
    "    return gene2tss\n",
    "\n",
    "def sum_peaks(peaks_plus, peaks_minus):\n",
    "    peaks_total = []\n",
    "    for p in peaks_plus:\n",
    "        peaks_total.append((p[1], '+', p[0]))\n",
    "    for p in peaks_minus:\n",
    "        peaks_total.append((p[1], '-', p[0]))\n",
    "    return peaks_total "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "min height: 5 & window size: 3\n",
      "n genes mapped in all replicates: 153 (early: 129, late: 153)\n",
      "median number of TSSs mapped: 12 (early: 2)\n",
      "minimum correspondence between replicates: 0.81\n",
      "\n",
      "min height: 5 & window size: 5\n",
      "n genes mapped in all replicates: 153 (early: 128, late: 153)\n",
      "median number of TSSs mapped: 10 (early: 2)\n",
      "minimum correspondence between replicates: 0.77\n",
      "\n",
      "min height: 5 & window size: 10\n",
      "n genes mapped in all replicates: 153 (early: 128, late: 153)\n",
      "median number of TSSs mapped: 8 (early: 2)\n",
      "minimum correspondence between replicates: 0.75\n",
      "\n",
      "min height: 10 & window size: 3\n",
      "n genes mapped in all replicates: 152 (early: 108, late: 152)\n",
      "median number of TSSs mapped: 8 (early: 2)\n",
      "minimum correspondence between replicates: 0.82\n",
      "\n",
      "min height: 10 & window size: 5\n",
      "n genes mapped in all replicates: 152 (early: 108, late: 152)\n",
      "median number of TSSs mapped: 7 (early: 1)\n",
      "minimum correspondence between replicates: 0.79\n",
      "\n",
      "min height: 10 & window size: 10\n",
      "n genes mapped in all replicates: 152 (early: 108, late: 152)\n",
      "median number of TSSs mapped: 6 (early: 1)\n",
      "minimum correspondence between replicates: 0.78\n",
      "\n",
      "min height: 20 & window size: 3\n",
      "n genes mapped in all replicates: 150 (early: 96, late: 150)\n",
      "median number of TSSs mapped: 6 (early: 1)\n",
      "minimum correspondence between replicates: 0.92\n",
      "\n",
      "min height: 20 & window size: 5\n",
      "n genes mapped in all replicates: 150 (early: 96, late: 150)\n",
      "median number of TSSs mapped: 5 (early: 1)\n",
      "minimum correspondence between replicates: 0.91\n",
      "\n",
      "min height: 20 & window size: 10\n",
      "n genes mapped in all replicates: 150 (early: 96, late: 150)\n",
      "median number of TSSs mapped: 4 (early: 1)\n",
      "minimum correspondence between replicates: 0.90\n",
      "\n",
      "min height: 30 & window size: 3\n",
      "n genes mapped in all replicates: 150 (early: 84, late: 150)\n",
      "median number of TSSs mapped: 5 (early: 1)\n",
      "minimum correspondence between replicates: 0.93\n",
      "\n",
      "min height: 30 & window size: 5\n",
      "n genes mapped in all replicates: 150 (early: 84, late: 150)\n",
      "median number of TSSs mapped: 4 (early: 1)\n",
      "minimum correspondence between replicates: 0.91\n",
      "\n",
      "min height: 30 & window size: 10\n",
      "n genes mapped in all replicates: 150 (early: 84, late: 150)\n",
      "median number of TSSs mapped: 4 (early: 1)\n",
      "minimum correspondence between replicates: 0.90\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## get max from a window\n",
    "for min_height in [5,10,20,30]:\n",
    "    for window_size in [3,5,10]:\n",
    "        peaks = {}\n",
    "        gene2tss = {}\n",
    "        \n",
    "        peaks['early R1 plus'] = get_peaks(early_R1_plus, window_size, min_height)\n",
    "        peaks['early R1 minus'] = get_peaks(early_R1_minus, window_size, min_height)\n",
    "        gene2tss['early R1'] = get_genes_starts(genes, sum_peaks(peaks['early R1 plus'], peaks['early R1 minus']))\n",
    "        \n",
    "        peaks['early R2 plus'] = get_peaks(early_R2_plus, window_size, min_height)\n",
    "        peaks['early R2 minus'] = get_peaks(early_R2_minus, window_size, min_height)\n",
    "        gene2tss['early R2'] = get_genes_starts(genes, sum_peaks(peaks['early R2 plus'], peaks['early R2 minus']))\n",
    "\n",
    "        peaks['late R1 plus'] = get_peaks(late_R1_plus, window_size, min_height)\n",
    "        peaks['late R1 minus'] = get_peaks(late_R1_minus, window_size, min_height)\n",
    "        gene2tss['late R1'] = get_genes_starts(genes, sum_peaks(peaks['late R1 plus'], peaks['late R1 minus']))\n",
    " \n",
    "        peaks['late R2 plus'] = get_peaks(late_R2_plus, window_size, min_height)\n",
    "        peaks['late R2 minus'] = get_peaks(late_R2_minus, window_size, min_height)\n",
    "        gene2tss['late R2'] = get_genes_starts(genes, sum_peaks(peaks['late R2 plus'], peaks['late R2 minus']))\n",
    "\n",
    "        print('min height:', min_height, '& window size:', window_size)\n",
    "        all_genes = set(); late_genes = set(); early_genes = set(); ntss = []; early_genes_ntss = []\n",
    "        for k in gene2tss:\n",
    "            all_genes.update(set(gene2tss[k].keys()))\n",
    "            if 'late' in k:\n",
    "                late_genes.update(set(gene2tss[k].keys()))\n",
    "            elif 'early' in k:\n",
    "                early_genes.update(set(gene2tss[k].keys()))\n",
    "                early_genes_ntss += [len(gene2tss[k][x]) for x in gene2tss[k]]\n",
    "            ntss += [len(gene2tss[k][x]) for x in gene2tss[k]]\n",
    "\n",
    "        print('n genes mapped in all replicates: %d (early: %d, late: %d)' % (len(all_genes), \n",
    "                                                                              len(early_genes), len(late_genes)))\n",
    "        print('median number of TSSs mapped: %d (early: %d)' % (numpy.median(ntss), \n",
    "                                                                   numpy.median(early_genes_ntss)))\n",
    "        correspondeces = [check_correspondence(peaks['early R1 plus'], peaks['early R2 plus']),\n",
    "                          check_correspondence(peaks['early R1 minus'], peaks['early R2 minus']),\n",
    "                          check_correspondence(peaks['late R1 plus'], peaks['late R2 plus']),\n",
    "                          check_correspondence(peaks['late R1 minus'], peaks['late R2 minus'])\n",
    "                         ]\n",
    "        print('minimum correspondence between replicates: %.2f' % min(correspondeces))  \n",
    "        print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAEChJREFUeJzt3X+s3Xddx/Hny44Os+HY2DSka+1ml4ViDIybDX+EECXSbpQSYrSNfwA2a/hRoyYmlGAM/AcYE12oLkWbgiEbZaK2oWQS4jKMla3DAa3N4FJGdu1CNwdTjHEO3v5xvoOz672n59xzzr09H56P5Oae8znn+/2+v/d77qvfvr/f+/2mqpAktevH1roASdJ0GfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxl2y1gUAXH311bV58+a1LkOSZspDDz30ZFVdc6H3rWnQJ9kB7NiyZQsnT55cy1IkaeYk+eYw71vT1k1VHauqvVdcccValiFJTbNHL0mNM+glqXEGvSQ1zqCXpMYZ9JLUuDUN+iQ7khx8+umn17IMSWqap1dKUuMuir+MHcfm/Z9e8bSPfuC2CVYiSRcne/SS1DiDXpIaZ9BLUuMMeklq3MSDPslrk3w+yZ1JXjvp+UuSRjNU0Cc5lOR8klOLxrcleSTJfJL93XAB3wVeCCxMtlxJ0qiG3aM/DGzrH0iyDjgAbAe2AruTbAU+X1XbgXcD759cqZKklRgq6KvqfuCpRcM3A/NVdbaqngHuBnZW1fe7178NXDqxSiVJKzLOH0xtAB7re74A3JLkzcDrgRcDH15u4iR7gb0AmzZtGqMMSdIg4wR9lhirqvoU8KkLTVxVB4GDAHNzczVGHZKkAcY562YB2Nj3/Frg3Cgz8KJmkjR94wT9g8ANSa5Lsh7YBRwdZQZe1EySpm/Y0yvvAk4ANyZZSLKnqp4F9gH3AmeAI1V1epSFu0cvSdM3VI++qnYvM34cOL7ShVfVMeDY3Nzc7SudhyRpMG88IkmN88YjktQ49+glqXHu0UtS47xMsSQ1ztaNJDXO1o0kNc7WjSQ1zqCXpMbZo5ekxtmjl6TG2bqRpMYZ9JLUOINekhrnwVhJapwHYyWpcbZuJKlxBr0kNc6gl6TGGfSS1DjPupGkxnnWjSQ1ztaNJDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJatxUgj7JZUkeSvKGacxfkjS8oYI+yaEk55OcWjS+LckjSeaT7O976d3AkUkWKklamWH36A8D2/oHkqwDDgDbga3A7iRbk7wO+FfgWxOsU5K0QpcM86aquj/J5kXDNwPzVXUWIMndwE7gcuAyeuH/30mOV9X3J1axJGkkQwX9MjYAj/U9XwBuqap9AEneCjy5XMgn2QvsBdi0adMYZUiSBhkn6LPEWP3gQdXhQRNX1cEkjwM71q9f/6ox6pAkDTDOWTcLwMa+59cC50aZgRc1k6TpGyfoHwRuSHJdkvXALuDoKDPwMsWSNH3Dnl55F3ACuDHJQpI9VfUssA+4FzgDHKmq06Ms3D16SZq+Yc+62b3M+HHg+EQrkiRNlHeYkqTGeYcpSWqce/SS1Dj36CWpcV6mWJIaZ+tGkhpn60aSGmfrRpIaZ9BLUuPs0UtS4+zRS1LjbN1IUuMMeklqnEEvSY3zYKwkNc6DsZLUOFs3ktQ4g16SGmfQS1LjDHpJapxn3UhS4zzrRpIaZ+tGkhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNW7iQZ/kZUnuTHJPkndMev6SpNEMFfRJDiU5n+TUovFtSR5JMp9kP0BVnamqtwO/DsxNvmRJ0iiG3aM/DGzrH0iyDjgAbAe2AruTbO1eeyPwj8DnJlapJGlFhgr6qrofeGrR8M3AfFWdrapngLuBnd37j1bVLwC/OcliJUmju2SMaTcAj/U9XwBuSfJa4M3ApcDx5SZOshfYC7Bp06YxypAkDTJO0GeJsaqq+4D7LjRxVR1M8jiwY/369a8aow5J0gDjnHWzAGzse34tcG6UGXhRM0mavnGC/kHghiTXJVkP7AKOjjIDL1MsSdM37OmVdwEngBuTLCTZU1XPAvuAe4EzwJGqOj3Kwt2jl6TpG6pHX1W7lxk/zoADrheSZAewY8uWLSudhSTpArzxiCQ1zmvdSFLjvGesJDXO1o0kNc7WjSQ1ztaNJDXO1o0kNc7WjSQ1zqCXpMbZo5ekxtmjl6TG2bqRpMYZ9JLUOHv0ktQ4e/SS1DhbN5LUOINekhpn0EtS4wx6SWqcQS9JjfP0SklqnKdXSlLjbN1IUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxk0l6JO8KclHkvxdkl+dxjIkScMZOuiTHEpyPsmpRePbkjySZD7JfoCq+tuquh14K/AbE61YkjSSUfboDwPb+geSrAMOANuBrcDuJFv73vIH3euSpDUydNBX1f3AU4uGbwbmq+psVT0D3A3sTM8Hgc9U1ReXml+SvUlOJjn5xBNPrLR+SdIFjNuj3wA81vd8oRv7beB1wK8leftSE1bVwaqaq6q5a665ZswyJEnLuWTM6bPEWFXVHcAdF5w42QHs2LJly5hlSJKWM+4e/QKwse/5tcC5YSf2omaSNH3jBv2DwA1JrkuyHtgFHB12Yi9TLEnTN8rplXcBJ4Abkywk2VNVzwL7gHuBM8CRqjo97Dzdo5ek6Ru6R19Vu5cZPw4cX8nC7dFL0vR54xFJapzXupGkxnnPWElqnK0bSWqcrRtJapytG0lqnK0bSWqcrRtJapytG0lqnK0bSWqcrRtJapxBL0mNM+glqXEejJWkxnkwVpIaZ+tGkhpn0EtS4wx6SWqcQS9JjTPoJalxnl4pSY3z9EpJapytG0lqnEEvSY0z6CWpcQa9JDXOoJekxk086JNcn+Qvk9wz6XlLkkZ3yTBvSnIIeANwvqp+tm98G/CnwDrgL6rqA1V1FtgzC0G/ef+nx5r+0Q/cNqFKJGl6ht2jPwxs6x9Isg44AGwHtgK7k2ydaHWSpLENFfRVdT/w1KLhm4H5qjpbVc8AdwM7J1yfJGlM4/ToNwCP9T1fADYkeUmSO4FXJnnPchMn2ZvkZJKTTzzxxBhlSJIGGapHv4wsMVZV9e/A2y80cVUdBA4CzM3N1Rh1SJIGGGePfgHY2Pf8WuDcKDPwomaSNH3jBP2DwA1JrkuyHtgFHB1lBl7UTJKmb6igT3IXcAK4MclCkj1V9SywD7gXOAMcqarToyzcPXpJmr6hevRVtXuZ8ePA8ZUuvKqOAcfm5uZuX+k8JEmDeeMRSWqcNx6RpMZ5UTNJatw459GPLckOYMeWLVvWsowVG+daOV4nR9JqsXUjSY2zdSNJjfOsG0lqnK0bSWqcrRtJapxn3awRz9iRtFps3UhS42zdSFLjDHpJapxBL0mN8zx6SWqcB2MlqXG2biSpcQa9JDXOoJekxhn0ktQ4g16SGue1brRqxrm+z1rxukJqgadXSlLjbN1IUuMMeklqnEEvSY0z6CWpcQa9JDVu4qdXJrkM+DPgGeC+qvr4pJchSRreUHv0SQ4lOZ/k1KLxbUkeSTKfZH83/Gbgnqq6HXjjhOuVJI1o2NbNYWBb/0CSdcABYDuwFdidZCtwLfBY97bvTaZMSdJKDRX0VXU/8NSi4ZuB+ao6W1XPAHcDO4EFemE/9PwlSdMzTo9+Az/cc4dewN8C3AF8OMltwLHlJk6yF9gLsGnTpjHK+NEzzqUE/JP+0cziZRvGNe5n5Eft8znuZ2Q11nmcoM8SY1VV/wW87UITV9VB4CDA3NxcjVGHJGmAcVorC8DGvufXAudGmYH3jJWk6Rsn6B8EbkhyXZL1wC7g6Cgz8KJmkjR9w55eeRdwArgxyUKSPVX1LLAPuBc4AxypqtOjLNw9ekmavqF69FW1e5nx48DxlS68qo4Bx+bm5m5f6TwkSYOt6emP7tFL0vR54xFJapx79JLUOPfoJalxqVr7v1VK8gTwzRVOfjXw5ATLWW3Wv7ZmvX6Y/XWw/pX76aq65kJvuiiCfhxJTlbV3FrXsVLWv7ZmvX6Y/XWw/unzomOS1DiDXpIa10LQH1zrAsZk/Wtr1uuH2V8H65+yme/RS5IGa2GPXpI0wEwH/TL3rL0oJHk0yVeSPJzkZDd2VZLPJvla9/3KbjxJ7ujW48tJbuqbz1u6938tyVumWO//uy/wJOtN8qru5zHfTbvU/QwmXf/7kvxbtw0eTnJr32vv6Wp5JMnr+8aX/Ex1V2n9Qrden+iu2DrJ+jcm+YckZ5KcTvI73fhMbIMB9c/ENkjywiQPJPlSV//7By0zyaXd8/nu9c0rXa9VUVUz+QWsA74OXA+sB74EbF3ruvrqexS4etHYh4D93eP9wAe7x7cCn6F3M5dXA1/oxq8Cznbfr+weXzmlel8D3AScmka9wAPAz3fTfAbYvgr1vw/4/SXeu7X7vFwKXNd9jtYN+kwBR4Bd3eM7gXdMuP6XAjd1j18EfLWrcya2wYD6Z2IbdD+Ty7vHLwC+0P1cl1wm8E7gzu7xLuATK12v1fia5T365e5ZezHbCXy0e/xR4E194x+rnn8GXpzkpcDrgc9W1VNV9W3gsyy6Sfuk1NL3BZ5Ivd1rP1FVJ6r32/CxvnlNs/7l7ATurqr/qapvAPP0Pk9Lfqa6Pd9fBu7ppu//WUyq/ser6ovd4/+kd+nvDczINhhQ/3Iuqm3Q/Ry/2z19QfdVA5bZv13uAX6lq3Gk9ZpU/Rcyy0G/1D1rB32wVlsBf5/kofTujwvwU1X1OPR+MYCf7MaXW5e1XsdJ1buhe7x4fDXs61obh55rezB6/S8BvlO9ezD0j09F1wZ4Jb29ypnbBovqhxnZBknWJXkYOE/vH8ivD1jmD+rsXn+6q/Gi/F2e5aBf8p61q17F8n6xqm4CtgPvSvKaAe9dbl0u1nUctd61Wo8/B34GeAXwOPDH3fhFW3+Sy4G/Bn63qv5j0FuXqWlN12GJ+mdmG1TV96rqFfRui3oz8LIBy7zo6h9kloN+7HvWTlNVneu+nwf+ht4H51vdf6Hpvp/v3r7cuqz1Ok6q3oXu8eLxqaqqb3W/vN8HPkJvG3CBOpcaf5Jea+SSReMTleQF9ELy41X1qW54ZrbBUvXP2jboav4OcB+9Hv1yy/xBnd3rV9BrHV6cv8urdTBg0l/07o51lt4Bj+cObrx8revqarsMeFHf43+i11v/I55/YO1D3ePbeP6BtQe68auAb9A7qHZl9/iqKda9mecfzJxYvfTuMfxqfngg8NZVqP+lfY9/j17vFODlPP+A2Vl6B8uW/UwBn+T5B+XeOeHaQ69v/ieLxmdiGwyofya2AXAN8OLu8Y8DnwfesNwygXfx/IOxR1a6XqvxtSoLmVrxvTMPvkqvl/beta6nr67ruw35JeD0c7XR6+F9Dvha9/25X8AAB7r1+Aow1zev36J3QGceeNsUa76L3n+t/5fe3seeSdYLzAGnumk+TPfHelOu/6+6+r5M78b1/aHz3q6WR+g7+2S5z1S3TR/o1uuTwKUTrv+X6P1X/svAw93XrbOyDQbUPxPbAPg54F+6Ok8BfzhomcALu+fz3evXr3S9VuPLv4yVpMbNco9ekjQEg16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMb9Hy15JBT6Ti1kAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "ax.hist(early_R2, 50)\n",
    "ax.set_yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
